### DISCUSSION TITLE
Enhanced Great Expectations Integration

### DISCUSSION THREAD

**Discussion by cicdw** [PrefectHQ/prefect#3558] [General]
In a joint effort with the great folks on the [Great Expectations](https://github.com/great-expectations/great_expectations) team, we are planning on enhancing our current integration with GE.  There are a few aspects to this work that should be very straightforward:

## Prior Art
@lauralorenz successfully researched and built an initial integration as a [Great Expectations checkpoint task](https://docs.prefect.io/api/latest/tasks/great_expectations.html) that is being used by a subset of our users.  This resulted in a still outstanding set of discussion points / future directions of work:
- High-level discussion: https://github.com/PrefectHQ/prefect/issues/2436
- Particular pain point about how GE config is handled: https://github.com/PrefectHQ/prefect/issues/3057

Because Prefect tasks can also be built on-the-fly using the `@task` decorator, there are likely other GE-specific uses and patterns out there that we might not be aware of.  

## Concrete Goals

In order to improve on the current state of affairs, we want to:
- [x] make any improvements to [the currently existing task](https://github.com/PrefectHQ/prefect/blob/master/src/prefect/tasks/great_expectations/checkpoints.py) that take recent GE configuration / data context changes into account (this "audit" should probably come from the GE side)
- [x] _possibly_ add other tasks to the task library that perform other GE functions other than just checkpointing, or at least document recommended patterns and gotchas
- [x] the Prefect team is already working on a new API for publishing task "artifacts", which includes arbitrary markdown that will be natively rendered in the UI; we want to improve the current checkpoint task to automatically publish the GE markdown report for UI visibility (**cc:** @joshmeek / @znicholasbrown )

## Notes

One of the biggest pain points from our earlier work was the differences in execution paradigms: it seemed GE was emphasizing an interactive workflow on a fixed machine, whereas Prefect is built on a deferred execution model in which Flows can be executed on any number of machines in the future, including a distributed network of machines (https://github.com/PrefectHQ/prefect/issues/3057 describes this well).  

Here is a minimal example of using the current GE task within a Flow: https://github.com/PrefectHQ/prefect/blob/master/examples/task_library/great_expectations/great_expectations_flow.py

This discussion can serve as a landing ground for ideas, action items, updates, etc. If we need to sync up on any sticking points, I'm happy to arrange for that as well.

Last but certainly not least, tagging @spbail for helping triage!

**Please comment on this discussion if you are a user using (or planning to use) Great Expectations + Prefect and let us know what you're doing, what's going well and what pain points you may be experiencing!**

**Comment by joshmeek** [comment]
Currently working on the artifacts API and I want to run some initial tests with posting markdown using a real world example. Looking through some of the great expectations docs/code now and one thing I'm wondering is what is the recommended way to turn an expectation suite into markdown? A quick look at the code revealed something like [render_multiple_validation_result_pages_markdown](https://github.com/great-expectations/great_expectations/blob/6ef3783eeee1f5f729b3f7c18372564e6fd5b1a9/great_expectations/render/page_renderer_util.py#L10) but it seems to be deprecated now with a different solution in the warning (which appears to be the contents of the function) and I'm wondering if this is the new recommended approach:

```python
validation_results_page_renderer = ValidationResultsPageRenderer(run_info_at_end=run_info_at_end)
rendered_document_content_list = validation_results_page_renderer.render_validation_operator_result(validation_operator_result=validation_operator_result)
return " ".join(DefaultMarkdownPageView().render(rendered_document_content_list))
```

cc @anthonyburdi

**Reply by anthonyburdi** [reply]
That is correct @joshmeek! 
We deprecated the utility function render_multiple_validation_result_pages_markdown() in favor of creating ValidationResultsPageRenderer.render_validation_operator_result() which can render `ValidationOperatorResult` objects by calling the render method for each `ExpectationSuiteValidationResult`. A `ValidationOperatorResult`can have multiple `ExpectationSuiteValidationResult`s, for example by running a checkpoint with multiple expectation suites.

**Comment by anthonyburdi** [comment]
The current [Great Expectations checkpoint task](https://github.com/PrefectHQ/prefect/blob/master/src/prefect/tasks/great_expectations/checkpoints.py#L22) looks great! To allow for a bit more flexible configuration of validation, we can potentially allow for the following:

1. **Allow for passing in an in-memory data context in lieu of reading from disk.** The Prefect task could take optional parameters for either the `context_root_dir` or an in-memory data `context` - and house the logic / error checking to make sure only a single one is used. Currently we do support an in memory DataContext (e.g. in Spark / Databricks notebooks), but we are currently in the process of enhancing the API and defaults to make this quite a bit less verbose. See [here](https://docs.greatexpectations.io/en/latest/guides/how_to_guides/configuring_data_contexts/how_to_instantiate_a_data_context_without_a_yml_file.html) and [here](https://docs.greatexpectations.io/en/latest/guides/how_to_guides/configuring_data_contexts/how_to_instantiate_a_data_context_on_a_databricks_spark_cluster.html) for our current implementation.

2. **Add additional configuration options for defining the validation.** Similar to 1., give the user the option to pass in one of the following (using optional parameters, with logic / error checking to make sure only one option is used) to define the validation: A. checkpoint, B. expectation_suite + batch_kwargs, C. assets_to_validate (e.g. `[{"suite_name": "my_suite", "batch_kwargs": {...}}, {"suite_name": "my_suite_2", "batch_kwargs": {...}}, ...]`)

**Comment by znicholasbrown** [comment]
Hi @anthonyburdi ! The UI team is picking up the rendering portion of this, where we're building a generic markdown renderer that we can use in a few different places - in this case for rendering task and flow run artifacts. Since we'll be using this for other purposes, I was hoping we could use GitHub-flavored markdown but whatever Great Expectations uses might be a better place to start; do you have documentation or some info we can use as a starting point for this? What flavor of markdown does GE use and is it extendible/configurable?

cc @ThatGalNatalie

**Reply by anthonyburdi** [reply]
Hi @znicholasbrown and @ThatGalNatalie!! Nice to meet you. We are using jinja templates to render our validation results to markdown, trying to adhere to GitHub-flavored markdown. You can take a look at all the files prefixed with `markdown_` in our [templates directory](https://github.com/great-expectations/great_expectations/tree/develop/great_expectations/render/view/templates). This reference doc may also be useful as it describes [customizing data docs](https://docs.greatexpectations.io/en/latest/reference/spare_parts/data_docs_reference.html?highlight=rendering#customizing-data-docs) which render to html but the concepts are similar for markdown.

Under the hood, we first render our validation results to a RenderedDocumentContent object (can be exported to json), and then consume it in our templates. The actual rendering happens in [`DefaultMarkdownPageView`](https://github.com/great-expectations/great_expectations/blob/develop/great_expectations/render/view/view.py#L456) where we do things like bold parameter names and add icons.

Hope that is helpful! Please don't hesitate to let me know if you have any questions.

**Comment by anthonyburdi** [comment]
Hi all! Just checking in to see if there is anything I can help with. Also, just so I can make sure to be available going forward: what is the timeline that you are working towards for this effort?

**Reply by joshmeek** [reply]
Hi @anthonyburdi the artifacts API mentioned in the original discussion has been built and released in one of the latest Prefect releases! Currently the API supports posting link and markdown artifacts to the backend during task runs in flows. The UI is still undergoing development to support the display of the artifacts however we expect that to be coming soon!

I was actually just playing around with attempting to generate a markdown data doc from one of the [tutorial datasets](https://github.com/superconductive/ge_tutorials/blob/main/ge_getting_started_tutorial/great_expectations/notebooks/pandas/validation_playground.ipynb) but instead of calling `context.open_data_docs()` I am using the direct method mentioned above that looks something like: 

```python
run_info_at_end = True
validation_results_page_renderer = ValidationResultsPageRenderer(run_info_at_end=run_info_at_end)
rendered_document_content_list = validation_results_page_renderer.render_validation_operator_result(validation_operator_result=results)
" ".join(DefaultMarkdownPageView().render(rendered_document_content_list))
```

Unfortunately it is raising an error on a missing jinja template and I'm wondering if you have seen this happen before or if I am doing something wrong in my attempts to render the markdown:

<details>
<summary>Traceback</summary>
<p>

```
Traceback (most recent call last):
  File "ge_artifacts.py", line 68, in <module>
    " ".join(DefaultMarkdownPageView().render(rendered_document_content_list))
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/great_expectations/render/view/view.py", line 478, in render
    for d in document
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/great_expectations/render/view/view.py", line 478, in <listcomp>
    for d in document
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/great_expectations/render/view/view.py", line 69, in render
    return t.render(document, **kwargs)
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/great_expectations/render/view/templates/markdown_validation_results_page.j2", line 6, in top-level template code
    {% include 'markdown_section.j2' %}
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/great_expectations/render/view/templates/markdown_section.j2", line 4, in top-level template code
    {% include 'markdown_component.j2' %}
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/great_expectations/render/view/templates/markdown_component.j2", line 16, in top-level template code
    {{ content_block | render_content_block }}
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/great_expectations/render/view/view.py", line 560, in render_content_block
    render_to_markdown=render_to_markdown,
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/great_expectations/render/view/view.py", line 191, in render_content_block
    jinja_context, content_block=content_block, index=index
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/great_expectations/render/view/templates/markdown_table.j2", line 15, in top-level template code
    {{ cell | render_content_block(content_block_id=content_block_id + '-content') }} {% if loop.last %} {% else %} | {% endif %}
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/great_expectations/render/view/view.py", line 560, in render_content_block
    render_to_markdown=render_to_markdown,
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/great_expectations/render/view/view.py", line 181, in render_content_block
    template = self._get_template(template=template_filename)
  File "/Users/josh/miniconda3/envs/prefect/lib/python3.7/site-packages/great_expectations/render/view/view.py", line 108, in _get_template
    template = env.get_template(template)
jinja2.exceptions.TemplateNotFound: markdown_content_block_container.j2
```

</p>
</details>

**Reply by znicholasbrown** [reply]
Hi @anthonyburdi - the API for this is in place (https://docs.prefect.io/api/latest/artifacts/artifacts.html#artifacts)! The UI has the initial renderer in place (https://github.com/PrefectHQ/ui/pull/390) and we plan to roll out the full exposure of the API in the next couple of weeks 😄

**Reply by anthonyburdi** [reply]
That is so great to hear @joshmeek @znicholasbrown!! Very exciting. I'll continue to keep my eye on this discussion.
FYI @joshmeek - I think you are correct that there is a missing jinja template! I've created a PR to add it here: https://github.com/great-expectations/great_expectations/pull/2063 but if you'd like to try it before it's released please feel free to install from the branch: [anthony/bugfix/add-markdown_content_block_container](https://github.com/great-expectations/great_expectations/tree/anthony/bugfix/add-markdown_content_block_container)

**Reply by spbail** [reply]
Wooo! I'm pretty excited about this, thanks for chipping away and thanks @anthonyburdi for the quick fix! 

Just to check in with the Prefect folks @cicdw - what are you planning in terms of timelines for the release and announcing the updated feature? I think it would make sense for us to have a how-to guide in the GE docs, and we'd love to do a blog post, too.

**Reply by cicdw** [reply]
Hey @spbail ! Official release target for the full feature (including UI) is Dec 4th, although we could easily begin the how-to guide and blog post anytime before then (screenshots will have to wait a little bit longer, but we already have the APIs so could write up code samples, etc. in the meantime).

**Reply by spbail** [reply]
That's fantastic, thanks @cicdw! I'd love to publish a blog post once the feature is out including some screenshots. Maybe we can get together once the UI is ready to do a walkthrough/demo together? Are you okay writing up the how-to on your end, and we'll link to it?

**Reply by cicdw** [reply]
Yup that sounds good to me!  I'll reach out next week 👍

**Reply by joshmeek** [reply]
@anthonyburdi the fix works flawlessly! Thanks for making that!

**Reply by spbail** [reply]
Hi folks, just wanted to check, is there a feature branch for the upcoming release? I'd love to check it out!

**Reply by joshmeek** [reply]
Hey @spbail @anthonyburdi I just pushed up a PR (#3753) to enhance the `RunGreatExpectationsCheckpoint` task in the task library w/ new kwargs and make use of markdown artifacts! (there's also a screenshot of the artifacts API in the UI there) I believe the UI branch will be available soon 🙌

**Reply by zanieb** [reply]
UI draft PR available at https://github.com/PrefectHQ/ui/pull/484

**Comment by spbail** [comment]
Howdy! Thanks @cicdw for taking the time to chat and demo today! I went over the code again and would suggest a few tweaks to make the input parameters (batch_kwargs+expectation suite, checkpoint, assets_to_validate) mutually exclusive, since a checkpoint name shouldn't be required when batch_kwargs+expectation suite are provided. Is it ok if I submit a PR to get started with this?

**Reply by cicdw** [reply]
Yes, absolutely!  The new kwargs haven't been released yet so shouldn't be breaking for anyone 👍

**Reply by spbail** [reply]
I've just opened a draft PR: #3766 

I don't have a local dev environment set up to actually test this yet, just wanted to get the code out to start the conversation. If anyone is able to commandeer the PR and run it, that would be super helpful - happy to make tweaks or provide input as needed.

**Reply by cicdw** [reply]
Sounds good, thanks @spbail!  I can make some tweaks on formatting, etc. Because the task name _was_ released in previous versions of Prefect I'll probably deprecate the old name and replace it with the new one instead of removing it in one fell swoop.

**Reply by spbail** [reply]
That sounds good re: deprecating the old name. Thanks!

**Comment by znicholasbrown** [comment]
A quick update on this for the GE folks (and anyone else who might be following along!): Artifacts are now released in the UI, meaning reports created with GE Tasks or more general markdown artifacts will now be visible on the Flow and Task Run pages in Cloud (and the latest version of Server).

For the GE folks, you can test the integration by signing up for a [Cloud account](https://cloud.prefect.io), and creating any flow with a GE task (or tasks). You'll be able to see the output of the GE task on the `Artifacts` tabs of either the flow run or task run pages.

![Screen Shot 2020-12-07 at 3 14 12 PM](https://user-images.githubusercontent.com/27291717/101400419-e69d7880-389e-11eb-97fb-277e9493aaa9.png)
![Screen Shot 2020-12-01 at 7 39 05 PM](https://user-images.githubusercontent.com/27291717/101400625-2cf2d780-389f-11eb-90ac-12413f74bad9.png)


Let me know if you have any questions!

cc: @spbail @anthonyburdi @joshmeek @cicdw

**Reply by nishuvr** [reply]
Hey how exactly did u create this report can u please share the code?

**Reply by melgo-tb** [reply]
Hey @znicholasbrown is the artifacts UI available only on the prefect Cloud UI or is it open sourced?

**Reply by zanieb** [reply]
Artifacts is available in Prefect 1 Cloud and Server.

Artifacts are on the roadmap for Prefect 2 and will likely be open source (but I cannot guarantee).
